### PR TITLE
1D line extraction from 2D images

### PR DESCRIPTION
**Pull request by McHaillet**
I worked on the 2D to 1D Fourier slice extraction, which is functional (I tried it for tttsa, see teamtomo/tttsa#19). 

One thing that I haven't figured out is a 90 degrees addition to the in plane rotations for the rotation matrices for this function, compared to the affine transform I used before ([link to the line of code](https://github.com/McHaillet/tttsa/blob/1b91814033a12fb346894d968a56f01256bec3a8/src/tttsa/optimizers.py#L99)). One of the two is interpreting it wrongly.

The dft_center also needed a fix to run on GPU, I can also put that in a separate PR if you are not sure yet about this one and want that bug fix in.

I have not added the inverse operation of 1D to 2D slice insertion. Should that be added here as well?

### CODE REVIEW

File: src/torch_fourier_slice/slice_extraction/_extract_central_slices_rfft_2d.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    rfft_shape = (freq_grid.shape[-2],)
+    output_shape = (*stack_shape, *rfft_shape)
+
+    # get (b, 2, 1) array of yx coordinates to rotate
+    if fftfreq_max is not None:
+        normed_grid = einops.reduce(freq_grid**2, "w yx -> w", reduction="sum") ** 0.5
+        freq_grid_mask = normed_grid <= fftfreq_max
+        valid_coords = freq_grid[freq_grid_mask, ...]  # (b, yx)
+    else:
+        valid_coords = einops.rearrange(freq_grid, "w yx -> (w) yx")
```
Comment (alisterburt) [inline src/torch_fourier_slice/slice_extraction/_extract_central_slices_rfft_2d.py:35]: this rearrange looks useless, just leftover from copying the 3d -> 2d impl. where it did something?

File: src/torch_fourier_slice/project.py [hunk 2]
```diff
@@ -67,3 +67,67 @@ def project_3d_to_2d(
...
+    pad: bool
+        Whether to pad the volume 2x with zeros to increase sampling rate in the DFT.
+    fftfreq_max: float | None
+        Maximum frequency (cycles per pixel) included in the projection.
+
+    Returns
+    -------
+    projections: torch.Tensor
+        `(..., d)` array of projection images.
+    """
```
Comment (alisterburt) [inline src/torch_fourier_slice/project.py:96]: few minor things in this docstring to tidy up

**Review by alisterburt — COMMENTED** [review]
awesome - really clean (but I would say that... it mirrors my own implementation 😆)

few minor notes but happy for this to be merged when you're ready

re: the 90 degree rotation I can't see anything obvious - the difference I would expect between conventions/implementations would be an inversion of rotation angle rather than a 90 degree shift, is this actually what is happening? does a zero degree projection produce an image projected along y as expected? what about 90 and x?

**Comment by alisterburt** [comment]
I don't mind about the bugfix

**Comment by alisterburt** [comment]
slice insertion can wait/come in a separate PR - no need to overload this one

File: src/torch_fourier_slice/slice_extraction/_extract_central_slices_rfft_2d.py (hunk 1 above)
Comment (McHaillet) [reply src/torch_fourier_slice/slice_extraction/_extract_central_slices_rfft_2d.py]: you're right! I misunderstood what this einops command did. Also the normed grid was useless for the 1D case :+1:

**Comment by McHaillet** [comment]
> awesome - really clean (but I would say that... it mirrors my own implementation 😆)

Haha, I was about to say, its mostly your implementation :p which is indeed clean!

> few minor notes but happy for this to be merged when you're ready

I updated those things!
 
> re: the 90 degree rotation I can't see anything obvious - the difference I would expect between conventions/implementations would be an inversion of rotation angle rather than a 90 degree shift, is this actually what is happening? does a zero degree projection produce an image projected along y as expected? what about 90 and x?

Double checked the projection at 0 and 90, and at 0 it is projected along the y-axis and at 90 along the x-axis. So that makes sense.

For the reconstruction, the tilt-axis is aligned with the y-axis, so to find the common line the projection from 2d to 1d needs to happen along the x-axis, i.e. the 90 addition makes sense. 

The real-space 2d to 1d projection I wrote before in tttsa was also projecting along the x-axis after the affine transform: https://github.com/teamtomo/tttsa/blob/ecaa1bc0a92d0f17cc80c51734ebe46e045ed860/src/tttsa/projection/project_real.py#L52

So in conclusion, it is correct like this. And, I need to add 90 degrees to project along the x-axis!

**Comment by McHaillet** [comment]
For my part ready to merge, but please let me know if you have other things that should be fixed.

**Comment by alisterburt** [comment]
Glad you got to the bottom of the 90 degree offset - it makes sense and everything matches expectations here

Awesome work as always!
